### PR TITLE
Update Ubuntu LTS version on Actions workflows

### PR DESCRIPTION
**Pull request by alexdesiqueira**
## Description

Updating Ubuntu LTS version on the GitHub Actions workflows to the most recent one — from 18.04/20.04 to 22.04, currently.

### CODE REVIEW

File: .github/workflows/circleci.yml
Comment (alexdesiqueira) [inline .github/workflows/circleci.yml:4]: If this is not `ubuntu-22.04`, it is [`ubuntu-2204`](https://circleci.com/developer/machine/image/ubuntu-2004) :slightly_smiling_face:

**Comment by mkcor** [comment]
How come wouldn't `ubuntu-latest` pick Ubuntu 22.04 LTS (since it happens to be the latest)? :thinking: 

Maybe we want to hard-code this anyway, so we know for sure which version is being run?

**Comment by alexdesiqueira** [comment]
> How come wouldn't ubuntu-latest pick Ubuntu 22.04 LTS (since it happens to be the latest)? thinking

@mkcor I think it's best for us to stick with the LTS versions; I'm kinda against using `ubuntu-latest` here since it will point to `22.10` in Oct, `23.04` in Apr '23, ... we'll have less control if things change too much from one point version to another.

> Maybe we want to hard-code this anyway, so we know for sure which version is being run?

That's the idea :slightly_smiling_face:

File: .github/workflows/circleci.yml
Comment (alexdesiqueira) [reply .github/workflows/circleci.yml:4]: I'm not entirely sure if there's some magic happening on GHA before pointing it to CircleCI, but we could try to "break it both ways"; if it doesn't work one way, it surely is the other one! :laughing:

**Review by mkcor — APPROVED** [review]

**Comment by jarrodmillman** [comment]
> I'm kinda against using `ubuntu-latest` here since it will point to `22.10` in Oct, `23.04` in Apr '23, ... we'll have less control if things change too much from one point version to another.

I am pretty sure that isn't how `ubuntu-latest` works. My understanding was that `ubuntu-latest` pointed to the latest LTS release of Ubuntu (or at least after its first point update to ensure extra stability). 

https://github.blog/changelog/2022-08-09-github-actions-ubuntu-22-04-is-now-generally-available-on-github-hosted-runners/

So if you use `ubuntu-latest`, you should always get the most recent LTS release (with the caveat that they normally wait until the LTS release has its first update to ensure that it is stable).

**Comment by hmaarrfk** [comment]
If you will remove 18.04, i would ensure to bump eneverything forward by 1 LTS. This ensures that we have the same mix of "old LTS" and new LTS.

However, Ubuntu 18.04 is still considered supported by Ubuntu.
For many, this LTS is only 2.5 years old, since they may have installed it in early 2020.

I personally think it is a little premature to remove it from the test matrix. Ubuntu, is quite "modern" compared to. CentOS or Redhat, and so it helps ensure that we are still compatible with it.

Are you removing it because of some pending deprecation?

I'm somewhat agnostic to removing latest. It was meant to help us catch incompatibilities with new computers earlier.

**Comment by jarrodmillman** [comment]
If you decide to continue using 18.04 for now, you may want to make an issue to remove it before December:
https://github.blog/changelog/2022-08-09-github-actions-the-ubuntu-18-04-actions-runner-image-is-being-deprecated-and-will-be-removed-by-12-1-22/

**Comment by hmaarrfk** [comment]
jarrod that is agood point. I didn't pick up on the motivation before.

In that case, I would do:

* 20.04 -> 22.04
* 18.04 -> 20.04

To ensure that the same number of "Old LTS" remain

**Comment by alexdesiqueira** [comment]
[TL;DR] Thank you for the discussion, @jarrodmillman @hmaarrfk. I think we should use `ubuntu-latest`, unless we have a _very_ good reason to use old LTS here. 

> I am pretty sure that isn't how ubuntu-latest works. For instance, ubuntu-latest currently is 20.04.4. In a month or so, it will be 22.04.1.
> (...)
> So if you use ubuntu-latest, you should always get the most recent LTS release (with the caveat that they normally wait until the LTS release has its first update to ensure that it is stable).

@jarrodmillman thanks for that! I didn't find that info before, and assumed `ubuntu-latest` would follow the... latest version. :slightly_smiling_face: I'd go towards your recommendation: we should point everything to `ubuntu-latest`, then.

> If you will remove 18.04, i would ensure to bump eneverything forward by 1 LTS. This ensures that we have the same mix of "old LTS" and new LTS.

@hmaarrfk I don't understand why we should have tests running that "old" software — or even 20.04, for that end. At some point you have a mix of really old packages and bleeding edge skimage, numpy, scipy... that could bring more headaches related to dependency solution than benefits of "supporting" that system. Could you explain your reason?

(A small side note: if we want something like that, it would be best to use Debian stable; more support, more robust base system, larger userbase (that includes Ubuntu). I don't know how easy Debian machines are as an option, though :slightly_smiling_face:)

> Ubuntu 18.04 is still considered supported by Ubuntu.
For many, this LTS is only 2.5 years old, since they may have installed it in early 2020.

I'd be interested to see how "many" are the many users you're referring to here. That release is "oldoldstable" (using Debian's terms), independently on when a possible user installed it. I searched for a number on how many users are according to each version, but couldn't find one.

> I personally think it is a little premature to remove it from the test matrix. Ubuntu, is quite "modern" compared to. CentOS or Redhat, and so it helps ensure that we are still compatible with it.
> I'm somewhat agnostic to removing latest. It was meant to help us catch incompatibilities with new computers earlier.

Ubuntu LTS is not the way to go if we're aiming for "new" computers. Also, CentOS/RH use a different package format, base system... a whole different beast. If we want to:
 1. provide support for newer packages: a machine for openSUSE Tumbleweed or Debian testing would suffice.
 2. test RPM packages: openSUSE Tumbleweed, or Fedora.

Fedora would be optimal to solve both 1. and 2. here.
Thanks again!

**Comment by hmaarrfk** [comment]
> @hmaarrfk I don't understand why we should have tests running that "old" software — or even 20.04, for that end. [...] Could you explain your reason?

My reason is quite simple. I consider myself as somebody running "bleading edge software" and I still found myself installing Ubuntu 20.04 a few weeks back due to a hardware incompatibility.

I don't think we need to be too quick to drop versions. manylinux2014 seems to be OK on Ubuntu 20.04. I don't think we need to push the envelope more than that. Is there a feature we *need* that is only available on 22.04?

My testing against "old" distributions, we are in a better place to help users that have errors when they show up.

**Comment by alexdesiqueira** [comment]
> My reason is quite simple. I consider myself as somebody running "bleading edge software" and I still found myself installing Ubuntu 20.04 a few weeks back due to a hardware incompatibility.

@hmaarrfk with all due respect, Mark, I don't see how anecdotal evidence is relevant for us to keep a specific version of (old :slightly_smiling_face:) software in our machinery. It's more likely that we have end users running "newer" versions of Ubuntu, as it will bother you once in a while to upgrade:

image

Personally, I've used every major (and maybe even minor!) Linux distros/BSD systems under the sun. I don't care for bleeding edge software, and Ubuntu 22.04 is far, far from it. If we're talking end users, I'd bet a cold non-alcoholic beverage that our largest share is running either Windows or macOS... :wink:

> I don't think we need to be too quick to drop versions. manylinux2014 seems to be OK on Ubuntu 20.04. I don't think we need to push the envelope more than that. Is there a feature we need that is only available on 22.04?

We don't _need_ anything, but I'd like to rewrite/improve/refactor our workflows soon, and it would be nice to have a more up-to-date machinery. This is just the start :slightly_smiling_face: 

> My testing against "old" distributions, we are in a better place to help users that have errors when they show up.

GHA workflows and the software we use here are not representative of real users systems, but you know this way better than me :slightly_smiling_face:

As I said, this:

> [TL;DR] Thank you for the discussion, @jarrodmillman @hmaarrfk. I think we should use ubuntu-latest, unless we have a very good reason to use old LTS here.

is what I'd prefer. I'll leave it open to you and @scikit-image/core to present your points and discuss it if you'd like. Since another maintainer reviewed and approved it already (thanks @mkcor!), I'll give y'all two weeks. Thanks!

**Comment by hmaarrfk** [comment]
I've given my thoughts. You've acknowledged them, you are free to move on with this PR as you wish. I am not "requiring" any changes, just suggesting.

**Review by jarrodmillman — APPROVED** [review]

**Comment by alexdesiqueira** [comment]
Thanks everyone!

**Pull request merged**
